### PR TITLE
Add TORCH_DCHECK macro that checks only in debug builds

### PR DESCRIPTION
**Pull request by bddppq**
Summary:
Follow up on discoveries/discussions in https://github.com/pytorch/pytorch/pull/30810

Mimic the `DCHECK` macro from https://github.com/pytorch/pytorch/blob/e5eb871/c10/util/logging_is_not_google_glog.h#L117-L125

With this change the perf gap is eliminated:

```
================================================================================
Program Output:
================================================================================
Run on (36 X 1601 MHz CPU s)
2019-12-12 20:12:13
-----------------------------------------------------------------
Benchmark                          Time           CPU Iterations
-----------------------------------------------------------------
BM_IntrusivePtrCtorDtor           23 ns         23 ns   30914703
BM_SharedPtrCtorDtor              27 ns         27 ns   25895944
BM_IntrusivePtrArray/16          503 ns        503 ns    1392139
BM_IntrusivePtrArray/32         1006 ns       1006 ns     695749
BM_IntrusivePtrArray/64         2013 ns       2013 ns     347714
BM_IntrusivePtrArray/128        4024 ns       4024 ns     173964
BM_IntrusivePtrArray/256        8047 ns       8047 ns      86994
BM_IntrusivePtrArray/512       16106 ns      16106 ns      43461
BM_IntrusivePtrArray/1024      32208 ns      32207 ns      21731
BM_IntrusivePtrArray/2048      64431 ns      64430 ns      10865
BM_IntrusivePtrArray/4096     128940 ns     128938 ns       5429
BM_SharedPtrArray/16             503 ns        503 ns    1392128
BM_SharedPtrArray/32            1006 ns       1006 ns     695940
BM_SharedPtrArray/64            2012 ns       2012 ns     347817
BM_SharedPtrArray/128           4024 ns       4023 ns     173927
BM_SharedPtrArray/256           8069 ns       8069 ns      86741
BM_SharedPtrArray/512          16143 ns      16142 ns      43357
BM_SharedPtrArray/1024         32283 ns      32283 ns      21685
BM_SharedPtrArray/2048         64718 ns      64717 ns      10817
BM_SharedPtrArray/4096        129469 ns     129466 ns       5407
================================================================================
```
```
================================================================================
Program Output:
================================================================================
Run on (80 X 2001 MHz CPU s)
2019-12-12 20:12:23
-----------------------------------------------------------------
Benchmark                          Time           CPU Iterations
-----------------------------------------------------------------
BM_IntrusivePtrCtorDtor           18 ns         18 ns   38630411
BM_SharedPtrCtorDtor              22 ns         22 ns   32356114
BM_IntrusivePtrArray/16          402 ns        402 ns    1739637
BM_IntrusivePtrArray/32          805 ns        805 ns     869818
BM_IntrusivePtrArray/64         1610 ns       1609 ns     434881
BM_IntrusivePtrArray/128        3218 ns       3218 ns     217437
BM_IntrusivePtrArray/256        6436 ns       6436 ns     108739
BM_IntrusivePtrArray/512       12882 ns      12882 ns      54356
BM_IntrusivePtrArray/1024      25763 ns      25763 ns      27177
BM_IntrusivePtrArray/2048      51532 ns      51531 ns      13590
BM_IntrusivePtrArray/4096     103091 ns     103091 ns       6778
BM_SharedPtrArray/16             402 ns        402 ns    1740165
BM_SharedPtrArray/32             804 ns        804 ns     869035
BM_SharedPtrArray/64            1610 ns       1610 ns     434975
BM_SharedPtrArray/128           3218 ns       3218 ns     217505
BM_SharedPtrArray/256           6457 ns       6457 ns     108510
BM_SharedPtrArray/512          12909 ns      12909 ns      54249
BM_SharedPtrArray/1024         25810 ns      25810 ns      27127
BM_SharedPtrArray/2048         51763 ns      51763 ns      13531
BM_SharedPtrArray/4096        103506 ns     103505 ns       6759
================================================================================
```

Test Plan:
buck test caffe2/c10/...
buck test @mode/opt caffe2/c10/...

Differential Revision: D18998243

### CODE REVIEW

**Comment by suo** [comment]
Hm, we should make sure OSS releases are compiled with NDEBUG defined (and that the CI is run with NDEBUG not defined).

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit d9d7eeb:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=d9d7eeb83d0edb09cf6597b64d72cbc0ab66946d).

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| CircleCI <sup>pytorch-linux-xenial-py3-clang5-android-ndk-r19c-gradle-build-x86_32</sup> | <sup>pytorch android gradle build only x86_32 (for PR)</sup> | New in PR |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Review by ezyang — CHANGES_REQUESTED** [review]
I think the naming here is very confusing. `TORCH_CHECK` is the name for a never compiled away test checking for user errors. `TORCH_INTERNAL_ASSERT` is the name for a test for internal logic error made by developer. So what's `TORCH_DCHECK`? If I didn't know anything about the glog naming convention, I'd expect it to be related to `TORCH_CHECK`; but it's not, it only should be used for `TORCH_INTERNAL_ASSERT`. If this were just Caffe2, it would be reasonable to defer to glog naming conventions, but that's not the situation here. I suggest a name like `TORCH_INTERNAL_ASSERT_DEBUG_ONLY` or `_HOTPATH`

**Comment by yinghai** [comment]
Why is `TORCH_DCHECK` confusing? Glog and its `CHECK`, `DCHECK` convention is widely used and known.

**Comment by ezyang** [comment]
I don't care if we follow the glog convention or not (though, it is true that I personally find the glog convention confusing for programmers who haven't programmed in glog before). What I do care about is being consistent.

`TORCH_CHECK` currently means, "Use this macro to validate user provided input. These tests may fail do to user input, if this occurs, raise an exception." `TORCH_INTERNAL_ASSERT` currently means, "Use this macro to assert internal invariants which should never be false. If this occurs, raise an exception"

In glog world, `CHECK` means "Immediately crash if a condition is not met; this macro is not compiled away in DEBUG builds. Use this macro in cases where it is better to crash than to cause a security vulnerability. `DCHECK` means "Immediately crash if a condition is not met; but remove this macro in DEBUG builds." Citation: https://chromium.googlesource.com/chromium/src/+/master/styleguide/c++/c++.md

`TORCH_CHECK` does something completely different from glog `CHECK`. What this PR is proposing to do is add a `TORCH_DCHECK` macro which does something like `DCHECK` (well, not actually; you haven't switched TORCH_DCHECK to crash the process, you're still just raising an exception). So now you are in this weird halfway state where half of the API follows glog semantics, and half the API doesn't. I think this is strictly worse than the world where we don't follow glog at all (what I am advocating for), or the world where everything follows glog (but then you have to do a mass rename on `TORCH_CHECK` again on the rest of the codebase.)

**Comment by yinghai** [comment]
> TORCH_CHECK currently means, "Use this macro to validate user provided input. These tests may fail do to user input, if this occurs, raise an exception." TORCH_INTERNAL_ASSERT currently means, "Use this macro to assert internal invariants which should never be false. If this occurs, raise an exception"

This is what your interpretation of those macros, which isn't necessarily intuitive to other developers either. I actually don't care about what the naming is. It's a hint and everybody has an opinion. As a developer, we have to look into the code to understand what these checks do instead of taking it by face value anyway. It appalls me that we have to spend so much time debating on a naming instead of doing benchmark, moving forward and actually making code faster.

**Comment by bddppq** [comment]
@ezyang “TORCH_DCHECK” simply means “TORCH_CHECK in debug mode only”. Of course there are other ways to hint the “debug mode only” part, like “_DEBUG_ONLY” suffix as you proposed. The reason I chose to use a single character “D” as the hint is because it has been used in several publicly well known libraries so kinda established as a convention already in many c++ developers’ mind (plus IMO “_DEBUG_ONLY” is toooo long to be ergonomic). Using “D” as hint doesn’t necessarily mean we need to have the same “CHECK” semantics as other libraries, that’s orthogonal.

**Comment by ezyang** [comment]
> This is what your interpretation of those macros, which isn't necessarily intuitive to other developers either.

It is not just my interpretation. Here is the documentation for the macros in the code:

```
// A utility macro to provide assert()-like functionality; that is, enforcement
// of internal invariants in code.  It supports an arbitrary number of extra
// arguments (evaluated only on failure), which will be printed in the assert
// failure message using operator<< (this is useful to print some variables
// which may be useful for debugging.)
//
// Usage:
//    TORCH_INTERNAL_ASSERT(should_be_true);
//    TORCH_INTERNAL_ASSERT(x == 0, "x = ", x);
//
// Assuming no bugs in PyTorch, the conditions tested by this macro should
// always be true; e.g., it should be possible to disable all of these
// conditions without changing observable user behavior.  If you would like to
// do error reporting for user input, please use TORCH_CHECK instead.
//
// NOTE: It is SAFE to use this macro in production code; on failure, this
// simply raises an exception, it does NOT unceremoniously quit the process
// (unlike assert()).
//
```

and

```
// A utility macro to make it easier to test for error conditions from user
// input.  Like TORCH_INTERNAL_ASSERT, it supports an arbitrary number of extra
// arguments (evaluated only on failure), which will be printed in the error
// message using operator<< (e.g., you can pass any object which has
// operator<< defined.  Most objects in PyTorch have these definitions!)
//
// Usage:
//    TORCH_CHECK(should_be_true); // A default error message will be provided
//                                 // in this case; but we recommend writing an
//                                 // explicit error message, as it is more
//                                 // user friendly.
//    TORCH_CHECK(x == 0, "Expected x to be 0, but got ", x);
//
// On failure, this macro will raise an exception.  If this exception propagates
// to Python, it will convert into a Python RuntimeError.
//
// NOTE: It is SAFE to use this macro in production code; on failure, this
// simply raises an exception, it does NOT unceremoniously quit the process
// (unlike CHECK() from glog.)
//
```

> It appalls me that we have to spend so much time debating on a naming instead of doing benchmark, moving forward and actually making code faster.

I'm not arguing about naming just for the sake of arguing about naming. In https://github.com/pytorch/pytorch/issues/20287 I reported a problem where ambiguous naming of the `AT_ASSERT` macro meant that many people were misusing debugging assert for user error checking. This kind of error is benign when you always compile with asserts enabled, but when you want to only check in debug build (as you're doing here), use of the wrong macro means that a user error check will be *silently omitted*. And chances are this won't be caught in testing, as our testing for "library correctly reports error condition when input is malformed" is very bad. It is *precisely* for this reason that getting the naming right is so important. I agree, in an ideal world every developer would have memorized exactly what each of these macros mean and would know what the right thing to use is. But lets be honest: most people are going to get their start just by copy pasting an occurrence of this macro from some existing code into their use case. An unambiguous name that clearly says what is going on will help both the code writer and the reviewer realize if they're doing it wrong.

>  The reason I chose to use a single character “D” as the hint is because it has been used in several publicly well known libraries so kinda established as a convention already in many c++ developers’ mind.  Using “D” as hint doesn’t necessarily mean we need to have the same “CHECK” semantics as other libraries, that’s orthogonal.

I don't think it's a good idea to rely on "conventions" like this. An example of this biting me is exactly #20287: I assumed that developers would know that "assert" means "don't use this for user error checking." I did not anticipate how many people would not know about this convention. It's also worth noting that many of our developers come from Python land (and are not "rugged" C++ developers). Let's try to make it easier for them to do the right thing!

> (plus IMO “_DEBUG_ONLY” is toooo long to be ergonomic). 

That's the point! If you're not sure if you should use `DCHECK` or `CHECK`, you should probably use CHECK. The point of a long unwieldly suffix is to discourage people from using it unless they know what they're doing.

**Comment by Yangqing** [comment]
Honestly, guys, why not someone stamp and merge lol. Code wins argument.

File: c10/test/util/exception_test.cpp
Comment (Yangqing) [inline c10/test/util/exception_test.cpp]: nit: #ifdef NDEBUG and switch the two branches?

**Review by Yangqing — APPROVED** [review]

File: c10/test/util/exception_test.cpp
Comment (bddppq) [reply c10/test/util/exception_test.cpp]: switched

**Comment by bddppq** [comment]
Since
1. I don't see any data that can back up the arguments of current naming confusing its users. All seems to be only based on subjective assumptions. 
2. The "users" of this macro will most likely be core devs, and from the reactions of core devs who have expressed their opinions in comments above, current choice of naming is more favorable.

In order to move this PR forward so that our users can benefit from the perf improvements, I'm going to land it as is. If anyone still insists on more discussions about the naming, feel free to create another PR to propose the change.

**Pull request closed without merging**
